doc/website/blog/2022-03-14-ent-hooks-examples.md

```diff
@@ -0,0 +1,430 @@
+---
+title: A few examples for Ent hooks  
+author: Yoni Davidson
+authorURL: "https://github.com/yonidavidson"
+authorImageURL: "https://avatars0.githubusercontent.com/u/5472778"
+authorTwitter: yonidavidson
+---
+Hooks are a popular subject in our blog posts
+[Building Observable Ent Applications with Prometheus](/blog/2021/08/12/building-observable-ent-application-with-prometheus)
+[Sync Changes to External Data Systems using Ent Hooks](/blog/2021/11/1/sync-to-external-data-systems-using-hooks)
+but their reactive nature makes them a tool that is usually affiliated with more advanced users.
+My goal in this blog post is to show you a few use cases where hooks can help you when building servers, and provide you with a few go-to recipes.
+These examples can be later reviewed in this [repo](https://github.com/yonidavidson/ent-hooks-examples).
+
+### So first, let us review in a few words what are hooks?
+
+[Hooks](https://entgo.io/docs/hooks) are a feature of Ent that allows adding custom logic before and after operations that change the data entities.
+
+A mutation is an operation that creates a change in the database.
+There are 5 types of mutations:
+1. Create.
+2. UpdateOne.
+3. Update.
+4. DeleteOne.
+5. Delete.
+
+Hooks are functions that get and return an [ent.Mutator](https://pkg.go.dev/entgo.io/ent#Mutator).
+They function similar to the popular [HTTP middleware pattern](https://github.com/go-chi/chi#middleware-handlers).
+
+```go
+package example
+
+import (
+	"context"
+
+	"entgo.io/ent"
+)
+
+func exampleHook() ent.Hook {
+	//use this to init your hook
+	return func(next ent.Mutator) ent.Mutator {
+		return ent.MutateFunc(func(ctx context.Context, m ent.Mutation) (ent.Value, error) {
+			// Do something before mutation.
+			v, err := next.Mutate(ctx, m)
+			if err != nil {
+				// Do something if error after mutation.
+			}
+			// Do something after mutation.
+			return v, err
+		})
+	}
+}
+```
+
+In Ent, there are two types of mutation hooks - schema hooks and runtime hooks. Schema hooks are mainly used for defining
+custom mutation logic on a specific entity type, for example, syncing entity creation to another system. Runtime hooks, on the other hand, are used
+to define more global logic for adding things like logging, metrics, tracing, etc.
+
+In this post, I will focus mainly on schemas hooks.
+
+### So why use them?
+
+In ent, we try and push the entity's logic down to the schema level, this allows us to define everything about an entity close to its definition. Entity properties could be:
+1. Valid values.
+2. Types.
+3. Allowed mutations.
+4. Privacy.
+5. Required Side effect (heavy compute for example).
+
+---
+Oftentimes, it’s more comfortable to define these properties in the RPC level, a controller provides a context and all the mutations
```

Comment:
Don't necessarily agree with this statement. It's definitely the common practice in the industry and Ent is different from other frameworks in its approach. 
Its huge benefit IMO isn't simplifying the management of code (reactive/event-based systems are harder to read and debug and reason about), but rather from the ability to compose higher-order systems such as generated backend implementations.